src/disjoint_mut.rs

```diff
@@ -644,39 +657,35 @@ impl From<RangeFull> for Bounds {
 /// `.index((start.., ..len))` to achieve the same.
 /// It's not as clear what it means initially, but we use this idiom so much
 /// I think it might be worth it for clarity through brevity.
-impl From<(RangeFrom<usize>, RangeTo<usize>)> for Bounds {
-    fn from((start, len): (RangeFrom<usize>, RangeTo<usize>)) -> Self {
-        Self {
-            range: start.start..start.start + len.end,
-        }
+impl SliceBounds for (RangeFrom<usize>, RangeTo<usize>) {
+    fn to_range(&self, len: usize) -> Range<usize> {
+        let (RangeFrom { start }, RangeTo { end: range_len }) = *self;
+        start..start + range_len
     }
 }
 
-pub trait SliceBounds: TranslateRange + Into<Bounds> + Clone + Debug {}
-
-impl SliceBounds for Range<usize> {}
-impl SliceBounds for RangeFrom<usize> {}
-impl SliceBounds for RangeInclusive<usize> {}
-impl SliceBounds for RangeTo<usize> {}
-impl SliceBounds for RangeToInclusive<usize> {}
-impl SliceBounds for RangeFull {}
-impl SliceBounds for (RangeFrom<usize>, RangeTo<usize>) {}
-
 impl<T> DisjointMutIndex<[T]> for usize {
     type Output = <[T] as Index<usize>>::Output;
 
+    #[inline(always)] // Need to inline to see bounds checks in order to potentially elide them.
     #[cfg_attr(debug_assertions, track_caller)]
     unsafe fn get_mut(self, slice: *mut [T]) -> *mut Self::Output {
         // SAFETY: The safety precondition for this trait method requires that
         // we can immutably dereference `slice`.
+        let index = self;
         let len = unsafe { (*slice).len() };
-        if self < len {
+        if index < len {
             // SAFETY: We have checked that `self` is less than the allocation
             // length therefore cannot overflow. `slice` is a valid pointer into
             // an allocation of sufficient length.
-            unsafe { (slice as *mut T).add(self) }
+            unsafe { (slice as *mut T).add(index) }
         } else {
-            panic!("index out of bounds: the len is {len} but the index is {self}");
+            #[inline(never)]
+            #[cfg_attr(debug_assertions, track_caller)]
+            fn out_of_bounds(index: usize, len: usize) -> Infallible {
+                panic!("index out of bounds: the len is {len} but the index is {index}")
+            }
+            match out_of_bounds(index, len) {} // Infallible
```

Comment:
This seems like an unusual pattern, is this modeled after something?

Reply:
No.  I was hoping the compiler would understand `Infallible` better but it didn't.  In `std` they just use `!` but that's unstable. `match`ing on `Infallible` (a variantless-`enum`) seemed like it worked, so I did that.  If there's a better way I can definitely switch to it.

Reply:
Turns out you actually can use `!` stably as long as it's only the return type, so I switched to that in [`10f5165..e25ab1c`](https://github.com/memorysafety/rav1d/compare/10f5165a2d60c136c929a4df64e33475a8c78c2d..e25ab1c181c7bf6103ba2815937939ae70b37a94).